src/clothing_mod.cpp

```diff
@@ -0,0 +1,183 @@
+#include "clothing_mod.h"
+
+#include <string>
+#include <set>
+
+#include "generic_factory.h"
+#include "item.h"
+#include "debug.h"
+
+namespace
+{
+
+generic_factory<clothing_mod> all_clothing_mods( "clothing mods" );
+
+} // namespace
+
+static std::map<clothing_mod_type, std::vector<clothing_mod>> clothing_mods_by_type;
+
+/** @relates string_id */
+template<>
+bool string_id<clothing_mod>::is_valid() const
+{
+    return all_clothing_mods.is_valid( *this );
+}
+
+/** @relates string_id */
+template<>
+const clothing_mod &string_id<clothing_mod>::obj() const
+{
+    return all_clothing_mods.obj( *this );
+}
+
+namespace io
+{
+
+static const std::map<std::string, clothing_mod_type> clothing_mod_type_map = {{
+        { "acid", clothing_mod_type_acid },
+        { "fire", clothing_mod_type_fire },
+        { "bash", clothing_mod_type_bash },
+        { "cut", clothing_mod_type_cut },
+        { "encumbrance", clothing_mod_type_encumbrance },
+        { "warmth", clothing_mod_type_warmth },
+        { "storage", clothing_mod_type_storage },
+        { "invalid", clothing_mod_type_invalid }
+    }
+};
+
+template<>
+clothing_mod_type string_to_enum<clothing_mod_type>( const std::string &data )
+{
+    auto iter = clothing_mod_type_map.find( data );
+
+    if( iter == clothing_mod_type_map.end() ) {
+        debugmsg( "Invalid mod type '%s'.", data.c_str() );
```

Comment:
```suggestion
        debugmsg( "Invalid mod type '%s'.", data );
```
Message formatting generally does not need the `c_str()` when the value is a `std::string`. It's needed for other types like `string_id`.

Reply:
fixed